gensim/models/word2vec.py

```diff
@@ -1204,6 +1212,28 @@ def prepare_vocab(self, hs, negative, wv, update=False, keep_raw_vocab=False, tr
         sample = sample or self.sample
         drop_total = drop_unique = 0
 
+        # if max_final_vocab is specified instead of min_count
+        # pick a min_count which satisfies max_final_vocab as well as possible
+        if self.max_final_vocab is not None:
+            sorted_vocab = sorted(self.raw_vocab.keys(), key=lambda word: self.raw_vocab[word], reverse=True)
+
+            if self.max_final_vocab < len(sorted_vocab):
+                calc_min_count = self.raw_vocab[sorted_vocab[self.max_final_vocab]] + 1
+            else:
+                calc_min_count = 1
+
+            if calc_min_count > min_count:
+                logger.info("effective_min_count was set to %d due to max_finavocab being set to %d" %
+                           (calc_min_count, self.max_final_vocab))
+                effective_min_count = calc_min_count
+            else:
+                logger.info("""specified min_count = %d is larger that min_count calculated
+                               by max_final_vocab = %d, using specified min_count""" % (min_count, calc_min_count))
+                effective_min_count = min_count
+        else:
+            logger.info("max_final_vocab is None. Setting effective_min_count to specified min_count")
+            effective_min_count = min_count
+
```

Comment:
The behavior/logic here in this block all correct, but could be much shorter, with fewer branches & log-lines – and thus clearer to read/maintain. For example, you could start by setting `calc_min_count=1`. Then, if a `max_final_vocab` is set, and then further if it's smaller than the sorted_vocab length, increase `calc_min_count` to the new required value. (That is: no `else` needed.) Then at the top level set `effective_min_count = max(calc_min_count, min_count)` – no comparisons, `else`, or multiple-alternative-assignments needed.) Finally, a single INFO log line showing `max_final_vocab`, 'calc_min_count`, `min_count`, and `effective_min_count` will reveal all that anyone would need to know about what min-count-handling occurred. 

Separately: (1) for logging our usual practice is to prefer the multiple-arguments method of passing values-to-be-formatted over directly `%`-operator formatting. (This alternative avoids every having to interpolate the string, if logging is not set at the given level.) For example, use `logger.info('something %d %d', some_value1, some_value2)` rather than `logger.info('something %d %d' % (some_value1, some_value2))`. 

(2) It'd be good to keep the effective value as a property on the model for later debugging/understanding - such as in `self.effective_min_count`.

Reply:
Thank you for the notes.
I apologize for the bad code and for this PR taking so long.
Will make changes.

Reply:
No apologies necessary, progress has been consistent! Often the full set of concerns/tradeoffs only becomes clear after trying initial approaches… and, noticing all the consistency issues, and related supporting work (like updating `load()` for backward-compatibility of model properties), becomes much easier, almost second nature, after more experience with such codebases.